clang/lib/AST/Type.cpp

```diff
@@ -94,6 +94,11 @@ bool Qualifiers::isTargetAddressSpaceSupersetOf(LangAS A, LangAS B,
          (A == LangAS::Default &&
           (B == LangAS::cuda_constant || B == LangAS::cuda_device ||
            B == LangAS::cuda_shared)) ||
+         // In HLSL, the this pointer for member functions is in the default
+         // address space. This causes problem if the structure is in
+         // hlsl_device. We want to allow casting from hlsl_device to default
+         // until a proper solution for that issue is found.
+         (A == LangAS::Default && B == LangAS::hlsl_device) ||
```

Comment:
@llvm-beanz I've added a hack to make member functions "work", but having the address space cast will cause other problems.

I could remove this, and say that structs in structured buffers does not work yet? We still have other issues related to the layout anyway.

Reply:
Structs in structured buffers are largely working for the DXIL backend right now, so I'd rather not break that. Is the "right" thing to do here to get the "this" pointer into the device address space when the struct is inside of a structured buffer?

Reply:
I don't think we have designed a "right" thing yet. I'm not sure of all of the gotcha yet. @Keenuts You have looked at this more. What are the options you've considered?

Reply:
That's fair. I suppose having the implicit cast as a stopgap might be the best short-term solution then. I'm assuming omitting this would mean that structs in structured buffers would be disallowed in all backends, correct?